### PR TITLE
Add implementation and test for calculate_nft_ownership_layer_puzzle_hash

### PR DESCRIPTION
**Pull request by greimela**
Plus some review comments

### CODE REVIEW

File: clsp/fusion_singleton.clsp [hunk 1]
```diff
@@ -39,26 +39,22 @@
       ; who am I?
       (list ASSERT_MY_COIN_ID singleton_coin_id)
       (list ASSERT_MY_PUZZLEHASH (calculate_full_puzzle_hash SINGLETON_MOD_HASH LAUNCHER_ID LAUNCHER_PUZZLE_HASH singleton_inner_puzzlehash))
-      (list ASSERT_MY_AMOUNT 1)
+      (list ASSERT_MY_AMOUNT 1) ; This feels unnecessary. And if you ever create such a singleton with amount > 1 it is locked forever.
```
Comment (trgarrett) [inline clsp/fusion_singleton.clsp:37]: Thanks.

File: clsp/fusion_singleton.clsp [hunk 2]
```diff
@@ -39,26 +39,22 @@
...
-  (defun announce_p2_releases (singleton_coin_id nft_ids_to_release release_puzzlehash conditions_so_far)
-      (if (l (r nft_ids_to_release))
-        (announce_p2_releases singleton_coin_id (r nft_ids_to_release) release_puzzlehash
-          (c 
-            (list CREATE_PUZZLE_ANNOUNCEMENT (sha256 singleton_coin_id (f (f nft_ids_to_release)) release_puzzlehash))
-            conditions_so_far
-          )
-        )
-        (c 
+  ; You can simplify those for loops a lot by turning them upside down:
```
Comment (trgarrett) [inline clsp/fusion_singleton.clsp]: I feel dense, but I'm not quite following here. Could you describe the mechanism a bit more?

File: clsp/utils.clib [hunk 3]
```diff
@@ -60,14 +60,14 @@
...
-  ;;;      (sha256 1 nft_ownership_layer_mod_hash)
-  ;;;      (sha256 1 new_owner)
-  ;;;      transfer_program_hash
-  ;;;      inner_puzzle_hash
-  ;;;    )
-  ;;;)
+  (defun-inline calculate_nft_ownership_layer_puzzle_hash (NFT_OWNERSHIP_LAYER_MOD_HASH current_owner transfer_program_hash inner_puzzle_hash)
+      (curry_hashes NFT_OWNERSHIP_LAYER_MOD_HASH
+        (sha256 1 NFT_OWNERSHIP_LAYER_MOD_HASH)
+        (sha256 1 current_owner)
```
Comment (trgarrett) [inline clsp/utils.clib:60]: Thanks. I like the current_owner terminology better. I was trying to base it on what the ownership layer called it, but I agree it didn't make sense.

Comment (greimela) [reply clsp/utils.clib:60]: When I look at `nft_ownership_layer.clsp` in the chia-blockchain repo, they call that parameter CURRENT_OWNER too

File: clsp/p2_fusion.clsp [hunk 4]
```diff
@@ -22,7 +22,7 @@
         (list
             ; prove the parentage of the supplied NFT coin matches the expected derived puzzle hash
             (list ASSERT_MY_PUZZLEHASH (calculate_full_puzzle_hash SINGLETON_MOD_HASH nft_launcher_id LAUNCHER_PUZZLE_HASH nft_inner_puzzle_hash))
-            (list ASSERT_MY_AMOUNT 1)
```
Comment (trgarrett) [inline clsp/p2_fusion.clsp:25]: If I prove the NFT is the one I desire by virtue of its full puzzlehash, is there any value in asserting amount for the NFT at all? I am hard-coding it some places in my Python driver to 1, but if we encounter non-1 values in the wild (which I do realize are possible), I could update the Python driver without any blockchain impacts.

Comment (trgarrett) [reply clsp/p2_fusion.clsp:25]: Ewww...another potential corner case maybe?

Would it be possible to send standard XCH to the extended puzzlehash and convince the p2 singleton that it was an NFT? Thoughts @greimela @yakuhito @BrandtH22

Comment (Yakuhito) [reply clsp/p2_fusion.clsp:25]: The launcher ids are curried in - so you must have a singleton with the NFT launcher id. This is safe IMO.

File: clsp/fusion_singleton.clsp [hunk 5]
```diff
@@ -31,149 +26,98 @@
...
       ; who am I?
       (list ASSERT_MY_COIN_ID singleton_coin_id)
-      (list ASSERT_MY_PUZZLEHASH (calculate_full_puzzle_hash SINGLETON_MOD_HASH LAUNCHER_ID LAUNCHER_PUZZLE_HASH singleton_inner_puzzlehash))
-      (list ASSERT_MY_AMOUNT 1)
+      (list ASSERT_MY_PUZZLEHASH (calculate_full_puzzle_hash SINGLETON_STRUCT singleton_inner_puzzlehash))
+      (list ASSERT_MY_AMOUNT 1) ; This feels unnecessary. And if you ever create such a singleton with amount > 1 it is locked forever.
 
       ; spend myself forward
-      (list CREATE_COIN singleton_inner_puzzlehash 1 (list LAUNCHER_ID)) ;; opt to keep details beyond launcher ID encoded in eve coin spend
+      (list CREATE_COIN singleton_inner_puzzlehash 1 (list (f (r SINGLETON_STRUCT)))) ;; opt to keep details beyond launcher ID encoded in eve coin spend
```
Comment (Yakuhito) [inline clsp/fusion_singleton.clsp:40]: is this needed? can always be found in the parent's puzzle reveal

Comment (greimela) [reply clsp/fusion_singleton.clsp:40]: The puzzle doesn't need it. @trgarrett, are you planning to use the hint for easier lookups?

**Comment by trgarrett** [comment]
Yes, I actually intend to update the hint with some information that helps
1.) find fusion puzzles, generally 2.) find what project is claiming them.
People could be a nuisance and put garbage hints on other coins, but the
set of coins that are correct will be at most the size of all coins with
those hints.

File: clsp/utils.clib (hunk 3 above)
Comment (trgarrett) [reply clsp/utils.clib:60]: Which branch are you on? I see new_owner a bunch of places.

File: clsp/fusion_singleton.clsp (hunk 5 above)
Comment (trgarrett) [reply clsp/fusion_singleton.clsp:40]: Yes, I actually intend to update the hint with some information that helps
1.) find fusion puzzles, generally 2.) find what project is claiming them.
People could be a nuisance and put garbage hints on other coins, but the
set of coins that are correct will be at most the size of all coins with
those hints.

File: clsp/fusion_singleton.clsp [hunk 6]
```diff
@@ -31,149 +26,98 @@
...
-              (sha256
-                  (f (f offer_wallet_asserts))
-                  (r (f offer_wallet_asserts))
-              )
-          ) 
-          conditions_so_far
+          (assert_p2_condition SINGLETON_MOD_HASH LAUNCHER_PUZZLE_HASH (f nfts_to_unlock) OFFER_MOD_HASH)
+          (c
+            (announce_p2_condition SINGLETON_MOD_HASH LAUNCHER_PUZZLE_HASH (f nfts_to_unlock) singleton_coin_id OFFER_MOD_HASH)
+            (compute_nft_unlock_conditions SINGLETON_MOD_HASH LAUNCHER_PUZZLE_HASH OFFER_MOD_HASH (r nfts_to_unlock) singleton_coin_id offer_nonce)
```
Comment (trgarrett) [inline clsp/fusion_singleton.clsp:101]: I'm a little baffled as to why this function takes a conditions_so_far but doesn't handle it in the recursive call. Can you help explain @greimela ?

Comment (greimela) [reply clsp/fusion_singleton.clsp:101]: ~~I'm confused, my local code looks different and I don't know why the remote code doesn't mirror it.~~

Ah, you fixed it already 😅 
Your changes look good, I'm sorry for the state my code was in...

**Pull request merged**